include/pybind11/attr.h

```diff
@@ -505,97 +442,128 @@ struct process_attribute<arg_v> : process_attribute_default<arg_v> {
+                                     + std::to_string(r.arginfo_index) + " "
+                                     + std::to_string(RecordType::nargs_pos_only));
         }
-        // It also can't follow a kw_only, but a static_assert in pybind11.h checks that
     }
 };
 
-/// Process a parent class attribute.  Single inheritance only (class_ itself already guarantees
-/// that)
+/// Process a parent class attribute.  Single inheritance only (class_ itself already
+/// guarantees that)
 template <typename T>
 struct process_attribute<T, enable_if_t<is_pyobject<T>::value>>
     : process_attribute_default<handle> {
-    static void init(const handle &h, type_record *r) { r->bases.append(h); }
+    template <typename RecordType>
+    static void init(const handle &h, RecordType &r) {
+        r.bases.append(h);
+    }
 };
 
 /// Process a parent class attribute (deprecated, does not support multiple inheritance)
 template <typename T>
 struct process_attribute<base<T>> : process_attribute_default<base<T>> {
-    static void init(const base<T> &, type_record *r) { r->add_base(typeid(T), nullptr); }
+    template <typename RecordType>
+    static void init(const base<T> &, RecordType &r) {
+        r.add_base(typeid(T), nullptr);
+    }
 };
 
 /// Process a multiple inheritance attribute
 template <>
 struct process_attribute<multiple_inheritance> : process_attribute_default<multiple_inheritance> {
-    static void init(const multiple_inheritance &, type_record *r) {
-        r->multiple_inheritance = true;
+    template <typename RecordType>
+    static void init(const multiple_inheritance &, RecordType &r) {
+        r.multiple_inheritance = true;
     }
 };
 
 template <>
 struct process_attribute<dynamic_attr> : process_attribute_default<dynamic_attr> {
-    static void init(const dynamic_attr &, type_record *r) { r->dynamic_attr = true; }
+    template <typename RecordType>
+    static void init(const dynamic_attr &, RecordType &r) {
+        r.dynamic_attr = true;
+    }
 };
 
 template <>
 struct process_attribute<custom_type_setup> {
-    static void init(const custom_type_setup &value, type_record *r) {
-        r->custom_type_setup_callback = value.value;
+    template <typename RecordType>
+    static void init(const custom_type_setup &value, RecordType &r) {
+        r.custom_type_setup_callback = value.value;
     }
 };
 
 template <>
 struct process_attribute<is_final> : process_attribute_default<is_final> {
-    static void init(const is_final &, type_record *r) { r->is_final = true; }
+    template <typename RecordType>
+    static void init(const is_final &, RecordType &r) {
+        r.is_final = true;
+    }
 };
 
 template <>
 struct process_attribute<buffer_protocol> : process_attribute_default<buffer_protocol> {
-    static void init(const buffer_protocol &, type_record *r) { r->buffer_protocol = true; }
+    template <typename RecordType>
```

Comment:
Wow, these are a lot of template types that are just ignored. Is there a more efficient way to do this that doesn't cause a lot of template initialization? Like making the 2nd arg a void pointer or something?